### PR TITLE
Sentry/tracing with Example from Documenation

### PR DESCRIPTION
**Pull request by thinkocapo**
## Done
Got the code to match the official Javascript example here:  
https://docs.sentry.io/product/performance/getting-started/
image
- This includes `performCheckoutOnServer()` function so a lot of details are abstracted away from the `checkout()` function. It reads a lot cleaner, and matches documentation.

- Sentry SDK's 'Profiler' renames the transaction

- /checkout response's data now appears in a span in the javascript Transaction

- pageload (get_tools) data now appears in a span in the javascript Tranasction

#### docker-compose
[js pageload /toolstore tx](https://sentry.io/organizations/testorg-az/discover/will-frontend-react:7881b1b944884d5886b4eff544f98686/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=event.type%3Atransaction&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1) 

[py pageload get_tools tx](https://sentry.io/organizations/testorg-az/discover/python-test:5bf2d43de785429684be38f749270546/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=event.type%3Atransaction&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1)

[js checkout tx](https://sentry.io/organizations/testorg-az/discover/will-frontend-react:ab239e1088f446e184226e0884ec35d9/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=event.type%3Atransaction&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1)

[py checkout tx](https://sentry.io/organizations/testorg-az/discover/python-test:b1d15af874374ee182202a7b7963c323/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=event.type%3Atransaction&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1)

[py checkout error](https://sentry.io/organizations/testorg-az/discover/python-test:fe6074bdc51f475592781f2b35acb5df/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&project=1428657&project=5260888&query=event.type%3Aerror&sort=-timestamp&statsPeriod=24h&widths=266&widths=-1&widths=-1&widths=-1&widths=-1)

[js tx error](https://sentry.io/organizations/testorg-az/discover/will-frontend-react:a15f1bd880994f16a2abfebed57e531d/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&project=1428657&project=5260888&query=event.type%3Aerror&sort=-timestamp&statsPeriod=24h&widths=266&widths=-1&widths=-1&widths=-1&widths=-1)

#### gcp

[js pageload tx](https://sentry.io/organizations/testorg-az/discover/will-frontend-react:de5dd508b9214c46a36684ca55be5ed7/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=event.type%3Atransaction&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1)

[py get_tools tx](https://sentry.io/organizations/testorg-az/discover/python-test:3fe73244dfe94c53a63f04da36209346/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=event.type%3Atransaction&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1)

[js checkout tx](https://sentry.io/organizations/testorg-az/discover/will-frontend-react:cd48fc3867cf4ca9a7b05b20c4b335eb/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=event.type%3Atransaction&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1)

[py checkout tx](https://sentry.io/organizations/testorg-az/discover/python-test:6c58797c11484158b8cfc3ecb289ea75/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=event.type%3Atransaction&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1)

[py checkout error](https://sentry.io/organizations/testorg-az/discover/python-test:ceef1ff208a24f088103a88650c67a56/)

[js checkout error](https://sentry.io/organizations/testorg-az/discover/will-frontend-react:e0cc86150b1d493ea0431685ad9aca15/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=tags%5Btrace%5D%3A2006ff42f5aa4b5fafa6178b9b5b4a40&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1)

**2 transactions + 2 errors united by 1 Trace id**
[link](https://sentry.io/organizations/testorg-az/discover/results/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=tags%5Btrace%5D%3A2006ff42f5aa4b5fafa6178b9b5b4a40&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1)
image

### CODE REVIEW

File: react/src/components/App.js [hunk 1]
```diff
@@ -122,28 +125,50 @@ class App extends Component {
...
+
+      if (!response.ok) {
+        throw new Error(response.status + " - " + (response.statusText || "INTERNAL SERVER ERROR"));
+      }
 
-    if (!response.ok) {
-      throw new Error(response.status + " - " + (response.statusText || "INTERNAL SERVER ERROR"));
+      this.setState({ success: true });
+    } finally {
+      span.finish();
```
Comment (thinkocapo) [inline react/src/components/App.js:169]: error needs to `throw` before you call `span.finish()` or else the Transaction will never finish

File: react/src/components/App.js [hunk 2]
```diff
@@ -122,28 +125,50 @@ class App extends Component {
   }
 
   async checkout() {
+    let span, transaction, response
```
Comment (thinkocapo) [inline react/src/components/App.js:128]: span.finish transaction.finish aren't accessible (scoping) in the 'finally block, unless you declare them out here.

File: react/src/components/App.js [hunk 3]
```diff
@@ -122,28 +125,50 @@ class App extends Component {
...
+      };
+
+      // old way
+      // Integrations.BrowserTracing.startIdleTransaction({ name: "checkout"});
+
+      // Was advised to do it this way (not in docs)
+      // Sentry.configureScope(scope => scope.setSpan(transaction));
+
+      transaction = Sentry.startTransaction({ name: "checkout" });
+      Sentry.getCurrentHub().configureScope(scope => scope.setSpan(transaction));
```
Comment (thinkocapo) [inline react/src/components/App.js:143]: need to put the Transaction on Scope in order for the Transaction to get associated to the Error

**Comment by thinkocapo** [comment]
now that we've added `transaction.finish` the png's for the tools are no longer in the transaction. This is because we call `transaction.finish` as soon as the response comes back, before the browser+react have a chance to render the images.

I'm okay with this.
https://sentry.io/organizations/testorg-az/discover/will-frontend-react:41e7ae2ab4554b5a8cf1ff7ff36ae681/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=event.type%3Atransaction&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1

**Comment by thinkocapo** [comment]
we now have a better transaction name:

before:
image

after:
image

**Comment by thinkocapo** [comment]
@ndmanvar let me know if you'd like the object to be something other than **httpResponseData** here. I think this is informative. Better than just writing, "data" or "result"
@idosun 
image

**Comment by thinkocapo** [comment]
@ndmanvar @idosun 
setting the tools data as a span on the Page Load transaction
```
const transaction = Sentry.getCurrentHub()
      .getScope()
      .getTransaction();
    if (transaction) {
      let span = transaction.startChild({
        data: { toolsData: tools },
        op: "setting tools",
        description: "Setting tools description",
      });
      // Do something
      span.finish();
    }
    this.props.setTools(tools)
```
image

File: react/src/components/App.js [hunk 4]
```diff
@@ -139,12 +148,36 @@ class App extends Component {
     }).catch((err) => { throw Error(err) });
 
     if (!response.ok) {
-      throw new Error(response.status + " - " + (response.statusText || "INTERNAL SERVER ERROR"));
+      Sentry.captureException(new Error(response.status + " - " + (response.statusText || "INTERNAL SERVER ERROR")))
```
Comment (thinkocapo) [inline react/src/components/App.js:152]: reviewed with @ndmanvar last week that if you don't use Sentry.captureException here, then the app comes to a halt, and the transaction will never finish (transaction.finish()) later on.

In a real app, devs would be trying to more gracefully handle their errors anyways.

This code is abstracted into `performCheckoutOnServer` the real focus (Performance Monitoring) is still in `checkout()`

File: react/src/components/App.js [hunk 5]
```diff
@@ -139,12 +147,35 @@ class App extends Component {
...
+      this.setState({ hasError: true, success: false });
     }
-
-    this.setState({ success: true });
     
-    return response;
+    return { httpResponseData: response.status + " - " + response.statusText }
```
Comment (thinkocapo) [inline react/src/components/App.js:154]: this value needs to be an object, not a string. If it's a string, then the 'httpResponseData' won't show and you'll have a warning in the UI saying, "invalid span type, need object"

**Review by dashed — APPROVED** [review]

**Review by ndmanvar — APPROVED** [review]

**Pull request merged**